src/java.desktop/share/classes/javax/swing/text/html/HTMLDocument.java

```diff
@@ -2473,7 +2501,7 @@ public HTMLReader(int offset, int popDepth, int pushDepth,
             tagMap.put(HTML.Tag.SMALL, ca);
             tagMap.put(HTML.Tag.SPAN, ca);
             tagMap.put(HTML.Tag.STRIKE, conv);
-            tagMap.put(HTML.Tag.S, ca);
+            tagMap.put(HTML.Tag.S, conv);
```

Comment:
Should we not update the spec
https://github.com/openjdk/jdk/blob/71b46c38a820319851cade2a15d64a657d4d7284/src/java.desktop/share/classes/javax/swing/text/html/HTMLDocument.java#L2314-L2319
to mention that it is now `ConvertAction `and not `CharacterAction`

Reply:
That's good question.

We can't do it because `ConvertAction` is not public. In fact, `B`, `FONT`, `I`, `STRIKE`, `SUB`, `SUP`, `U` use `ConvertAction` instead of the specified `CharacterAction`.

The bigger problem is that `ConvertAction` does not extend `CharacterAction`, which means the implementation is different from what is specified. It is the result of [JDK-4171509](https://bugs.openjdk.org/browse/JDK-4171509).

There are a few more inconsistencies between the table and the real implementation. There also references to non-public action classes, such as `TitleAction`, `LinkAction`…

Reply:
Probably we can do away with those tags which mention internal class OR mention the public action instead, like TagAction instead of ConvertAction
HiddenAction instead of TitleAction etc

Reply:
I agree we should do something about it. I think `ConvertAction` can be subclass of `CharacterAction`, then the listed actions in the table will be more inline with the implementation.

Reply:
I've submitted [JDK-8325620](https://bugs.openjdk.org/browse/JDK-8325620): HTMLReader uses ConvertAction instead of specified CharacterAction for `<b>`, `<i>`, `<u>`

Reply:
> HiddenAction instead of TitleAction etc

I've submitted [JDK-8325623](https://bugs.openjdk.org/browse/JDK-8325623): HTMLReader refers to non-public classes as tag handlers

I decided to submit separate bugs because `ConvertAction` would likely have some source code changes. However, both bugs can be addressed as one changeset.